flanksource/karina
# feat: add Keptn production runtime

### Description

This PR includes 3 changes:
- Change manifest/upstream structure to allow more flexibility in defining upstream templates:
  - Support generating template files from helm chart (to support production runtime that uses helm chart installation method e.g. Keptn).
  - Each directory under `manifest/upstream` represents a workload/production-runtime template with the following structure
    ```
    upstream
    ├── argo-rollouts <-- (argo-rollouts production runtime)
    │   ├── base <-- (optional, if presents, `kustomize build` will be run in this dir to generate manifest/argo-rollouts.yaml)
    │   │   └── kustomization.yaml
    │   └── crd <-- (optional, if presents, `kustomize build` will be run in this dir to generate manifest/crd/argo-rollouts.yaml)
    │       └── kustomization.yaml
    └── keptn
        ├── base
        │   ├── configuration-service-patch.yaml
        │   ├── kustomization.yaml
        │   ├── mongodb-datastore-patch.yaml
        │   └── mongodb.yaml
        └── helm <-- (optional, if presents, chart.yaml and values.yaml will be use to generate manifests for the helm chart defined in chart.yaml)
            ├── chart.yaml
            ├── generated_templates <-- the generated templates from the helm chart will be output to this dir - git-ignored
            ├── values.yaml
    ```
  - `make` command will automatically build/re-build manifest/crd if there are dependency changes detected. For example, if there is a change in `keptn/helm/values.yaml`, `make` will rebuild the helm chart template -> rebuild `keptn/base` kustomization project and generated a new `manifest/keptn.yaml`
- Add a new karina template (`db.flanksource.com/v1/MongoDB`) that abstracts away a standard deployment of MongoDB (later we can switch out this deployment implementation with a more robust version)
- Add Keptn as a new Production Runtime, deployed in `platform-system` namespace.

Note: there are some TODO notes in the code to keep in mind what needs to be updated in the future when we implement a MongoDB Operator.

### Dependencies

- https://github.com/flanksource/kommons/pull/19

### Breaking Change


### Testing Notes

**What I did:**
**To test Keptn Production Runtime**
- Create a `cluster-config.yaml` (mind the `keptn` and `templateOperator` field)
```yaml
domain: 127.0.0.1.nip.io
name: test-cluster
calico:
  version: v3.8.2
kubernetes:
  version: v1.18.6
  masterIP: localhost
  containerRuntime: containerd
ca:
  cert: .certs/root-ca.crt
  privateKey: .certs/root-ca.key
  password: foobar
ingressCA:
  cert: .certs/ingress-ca.crt
  privateKey: .certs/ingress-ca.key
  password: foobar
keptn:
  version: v0.7.3
templateOperator:
  version: v0.1.10
```
- Provision with `karina provision kind-cluster -c test-cluster.yml`
- Deploy production runtime with `karina deploy phases --crds --calico --base --template-operator --keptn -c test-cluster.yml`
- Then run the test phase with `karina test phases --keptn -c test-cluster.yml` to verify that Keptn's Deployment objects are up and running.

**To test the new `manifest/Makefile`**
- Go to `manifest`
- Run `make` to update out-of-date templates, the command should exit without any error
- Run `make` again -> You should see `make: Nothing to be done for `build'.`
- Now go to `upstream/keptn/helm/values.yaml` add this line:
```yaml
control-plane:
  configurationService:
    storage: 500Mi
```
- Now run `make` in `manifest` again -> You should see the templates of the helm chart is re-generated -> then the `manifests/keptn/yaml` is updated (the make command runs `kustomize build` in `upstream/keptn/base` to generate it). You can also check your git diff to see what was changed.

**How you can replicate my testing:**
Same

### **Links**

Closes https://github.com/flanksource/karina/issues/654

Related to https://github.com/flanksource/karina/issues/658
